src/coprocessor/dag/rpn_expr/impl_compare.rs

```diff
@@ -0,0 +1,262 @@
+
+impl<C: Comparer> Copy for RpnFnCompare<C> {}
+
+impl<C: Comparer> Clone for RpnFnCompare<C> {
+    #[inline]
+    fn clone(&self) -> Self {
+        Self::new()
+    }
+}
+
+unsafe impl<C: Comparer> Send for RpnFnCompare<C> {}
+
+unsafe impl<C: Comparer> Sync for RpnFnCompare<C> {}
+
+// ======
+
+impl_template_fn! { 2 arg @ RpnFnCompare<C>, C: Comparer }
+
+pub trait Comparer: 'static {
+    type T: Evaluable;
+
+    fn compare(lhs: &Option<Self::T>, rhs: &Option<Self::T>) -> Result<Option<i64>>;
+}
+
+pub struct RealComparer<F: CmpOp> {
+    _phantom_f: std::marker::PhantomData<F>,
+}
+
+impl<F: CmpOp> Comparer for RealComparer<F> {
+    type T = Real;
+
+    #[inline]
+    fn compare(lhs: &Option<Real>, rhs: &Option<Real>) -> Result<Option<i64>> {
+        match (lhs, rhs) {
+            (None, None) => Ok(F::compare_null()),
+            (None, _) | (_, None) => Ok(F::compare_partial_null()),
+            (Some(lhs), Some(rhs)) => lhs
+                .partial_cmp(rhs)
+                // FIXME: It is wired to be a codec error.
+                // FIXME: This should never happen because special numbers like NaN and Inf are not
+                // allowed at all.
+                .ok_or_else(|| {
+                    Error::from(codec::Error::InvalidDataType(format!(
+                        "{} and {} can't be compared",
+                        lhs, rhs
+                    )))
+                })
+                .map(|v| Some(F::compare_order(v) as i64)),
+        }
+    }
+}
+
+pub trait CmpOp: 'static {
+    #[inline]
+    fn compare_null() -> Option<i64> {
+        None
+    }
+
+    #[inline]
+    fn compare_partial_null() -> Option<i64> {
+        None
+    }
+
+    fn compare_order(ordering: std::cmp::Ordering) -> bool;
+}
+
+pub struct CmpOpLT;
+
+impl CmpOp for CmpOpLT {
+    #[inline]
+    fn compare_order(ordering: Ordering) -> bool {
+        ordering == Ordering::Less
+    }
+}
+
+pub struct CmpOpLE;
+
+impl CmpOp for CmpOpLE {
+    #[inline]
+    fn compare_order(ordering: Ordering) -> bool {
+        ordering != Ordering::Greater
+    }
+}
+
+pub struct CmpOpGT;
+
+impl CmpOp for CmpOpGT {
+    #[inline]
+    fn compare_order(ordering: Ordering) -> bool {
+        ordering == Ordering::Greater
+    }
+}
+
+pub struct CmpOpGE;
+
+impl CmpOp for CmpOpGE {
+    #[inline]
+    fn compare_order(ordering: Ordering) -> bool {
+        ordering != Ordering::Less
+    }
+}
+
+pub struct CmpOpNE;
+
+impl CmpOp for CmpOpNE {
+    #[inline]
+    fn compare_order(ordering: Ordering) -> bool {
+        ordering != Ordering::Equal
+    }
+}
+
+pub struct CmpOpEQ;
+
+impl CmpOp for CmpOpEQ {
+    #[inline]
+    fn compare_order(ordering: Ordering) -> bool {
+        ordering == Ordering::Equal
+    }
+}
+
+pub struct CmpOpNullEQ;
+
+impl CmpOp for CmpOpNullEQ {
+    #[inline]
+    fn compare_null() -> Option<i64> {
+        Some(1)
+    }
+
+    #[inline]
+    fn compare_partial_null() -> Option<i64> {
+        Some(0)
```

Comment:
`null == 10` will return `Some(0)`, but it should return `None`
```
MySQL(127.0.0.1:3306)> create table realeq (a float, b float)
MySQL(127.0.0.1:3306)> insert into realeq values (null, 10), (null, null), (10, null), (10, 10), (0, 0)
MySQL(127.0.0.1:3306)> select a, b, a = b from test.realeq
+-----------------+
| a  | b  | a = b |
+-----------------+
|    | 10 |       |
|    |    |       |
| 10 |    |       |
| 10 | 10 | 1     |
| 0  | 0  | 1     |
+-----------------+
5 row in set (0.002 sec)

Reply:
This would describe `CmpOpEQ`

Reply:
in null equal, null equals to null

Reply:
see `builtinNullEQRealSig` in `expression/builtin_compare.go` for more detail